lib/AST/GenericSignatureBuilder.cpp

```diff
@@ -4165,17 +4228,27 @@ static ConstraintResult visitInherited(
   };
 
   // Visit all of the inherited types.
-  auto typeDecl = decl.dyn_cast<TypeDecl *>();
-  auto extDecl = decl.dyn_cast<ExtensionDecl *>();
   ArrayRef<TypeLoc> inheritedTypes = typeDecl ? typeDecl->getInherited()
                                               : extDecl->getInherited();
   for (unsigned index : indices(inheritedTypes)) {
     Type inheritedType = typeDecl ? typeDecl->getInheritedType(index)
                                   : extDecl->getInheritedType(index);
     if (!inheritedType) continue;
 
+    // If the inherited type is 'Any', allow it to pass through verbatim, as
+    // GSB diagnoses redundant 'Any' constraints. The reason we do this check
+    // here instead of inside visitInherited is in order to avoid diagnosing on
+    // constraints such as ': Any & P'. This is because:
+    // - 1: It preserves parity with the behaviour for protocol composition
+    // constraints that appear in 'where' clauses (which aren't decomposed until
+    // after the redundant 'Any' diagnostic).
+    // - 2: The diagnosis of 'Any & ...' protocol composititions is better done
+    // as a separate broader diagnostic that warns on the usage of such
+    // compositions anywhere, not just in constraints.
+    bool skipDecomposition = inheritedType->isEqual(ctx.TheAnyType);
+
     const auto &inherited = inheritedTypes[index];
-    visitInherited(inheritedType, inherited.getTypeRepr());
+    visitInherited(inheritedType, inherited.getTypeRepr(), skipDecomposition);
   }
```

Comment:
Thinking about it, why do we decompose protocol compositions that appear as inherited type requirements earlier than requirements in a `where` clause? Would it be possible to eliminate `visitInherited` and instead do a loop over the inherited types, calling `addTypeRequirement` for each one (I tried, and it didn't break any tests)? That would allow us to get rid of this somewhat ugly special case for `Any` and generally simplify the code a bit.